Frontend_Web_Development_React_Angular_Vue/React/Week 2/2.1_Comparing Functional and Class Components.md

```diff
@@ -0,0 +1,288 @@
+
+> An example would be shown soon, shortly
+
+![React Hooks Logo](https://spycoding.com/wp-content/uploads/2020/10/react_hooks.png)
+
+## What are React Hooks?
+
+Hooks is an additional API that is coming in the limelight with functional components. It is a blazing fire amongst the React ecosystem since the `v16.8` public release.
+
+React Hooks are in-built functions that allow React developers to use state and lifecycle methods inside functional components, they also work together with existing code, so they can easily be adopted into a codebase. The way Hooks were pitched to the public was that they allow developers to use state in functional components but under the hood, Hooks are much more powerful than that. They allow React Developers to enjoy the following benefits:
+
+-   Improved code reuse;
+-   Better code composition;
+-   Better defaults;
+-   Sharing non-visual logic with the use of custom hooks;
+-   Flexibility in moving up and down the `components` tree.
+
+With React Hooks, developers get the power to use functional components for almost everything they need to do from just rendering UI to also handling state and also logic — which is pretty neat.
+
+**To put things simple. Consider Hooks API provide a state to functional components (as they are stateless, being non-object oriented, not wrapped in a class).**
+
+Well, by now you must choose your way, between **class** and **functional** components approach.
+
+Coming to the example as told above, I would like to implement a **button clicked** state storage using both the approaches. This shall enable you to choose as per your comfort.
+
+```jsx
+import React from "react";
+
+class App extends React.Component {
+	constructor(props) {
+		super(props); // Inherited child classes should call parent's constructor.
+		this.state = {
+			buttonClicked: false,
+			clickedTimes: 0,
+			avgClickDuration: 0,
+		};
+	}
+	// render() prints the values in the state
+	render() {
+		let val = ``;
+		for (const key in this.state) val += `${key} : ${this.state[key]}, `;
+		return <div>{val}</div>;
+	}
+}
+
+export default App;
+```
+
+In the class context above,
+
+>     this.state = {
+>     			buttonClicked: false,
+>     			clickedTimes: 0,
+>     			avgClickDuration: 0,
+>     };
+
+is the definition of the state. It is an object [key-value] pairs.
+
+You must be puzzled seeing so much code at once, but to comfort you, the above approach gets much more simple when you use functional approach. Let me show you how,
+
+```jsx
+const ButtonClicked = () => {
+	const [buttonClicked, setButtonClicked] = useState({
+		// Default State
+		buttonClicked: false,
+		clickedTimes: 0,
+		avgClickDuration: 0,
+	});
+};
+export default ButtonClicked;
+```
```

Comment:
I cannot understand this. Is it supposed to be the functional component alternative of that. I was just wondering cause this does not render anything. Even if the render is same you should write it at least.

Reply:
Yes, this is the functional component based approach of the above class component approach. Although I had said it, but was less emphasized (not being bold)

Now I have fixed it to highlight properly.